File: ts-sdk/src/layers/sec.ts
import { BaseRecord, LogRecord } from '../core/record';
import { Layer } from '../core/enums';
import { Logger } from '../core/logger';
import { Level } from '../core/enums';

export interface SecRecordProps extends LogRecord {
  readonly actor_id: string | null;
  readonly action: string;
  readonly cve: string | null;
}

export class SecRecord extends BaseRecord implements SecRecordProps {
  readonly actor_id: string | null;
  readonly action: string;
  readonly cve: string | null;

  constructor(data: SecRecordProps) {
    super(data);
    this.actor_id = data.actor_id || null;
    this.action = data.action || 'UNKNOWN';
    this.cve = data.cve || null;
    Object.freeze(this);
  }

  toDict(): Record<string, any> {
    return { ...this, layer: Layer.SECURITY };
  }

  toCEF(): string {
    return `CEF:0|mylogger|ts-sdk|1.0|${this.action}|${this.message}|${this.level}|actor=${this.actor_id}`;
  }
}

export class SecLogger extends Logger<SecRecord> {
  createRecord(message: string, level: Level, ctx: Record<string, any>): SecRecord {
    const base = this.getBaseProps(message, level, Layer.SECURITY, ctx);
    return new SecRecord({
      ...base,
      actor_id: ctx.actor_id || null,
      action: ctx.action || 'UNKNOWN',
      cve: ctx.cve || null,
    });
  }
}

File: py-sdk/mylogger/layers/sec.py
from dataclasses import dataclass
from typing import Optional, Dict, Any
from ..core.record import LogRecord, Layer, Level
from ..core.logger import Logger

@dataclass(frozen=True)
class SecRecord(LogRecord):
    actor_id: Optional[str] = None
    action: str = "UNKNOWN"
    cve: Optional[str] = None

    def to_cef(self) -> str:
        # Simplified Common Event Format
        return f"CEF:0|mylogger|sec-sdk|1.0|{self.action}|{self.message}|{self.level}|actor={self.actor_id}"

class SecLogger(Logger[SecRecord]):
    def create_record(self, msg: str, level: Level, **ctx) -> SecRecord:
        base = self._get_base_props(msg, level, Layer.SECURITY, **ctx)
        return SecRecord(
            **base,
            actor_id=ctx.get("actor_id"),
            action=ctx.get("action", "UNKNOWN"),
            cve=ctx.get("cve")
        )

    def handle(self, record: SecRecord) -> None:
        # Security records are NEVER sampled and often written synchronously
        for handler in self.handlers:
            handler.handle(record)

    def should_handle(self, record: SecRecord) -> bool:
        return True # Security events always handled
